### PR TITLE
fix: addition and standardization of docstrings

### PR DESCRIPTION
**Pull request by zeqri**
#### Reference Issues/PRs
#164 
#### What does this implement/fix? Explain your changes.
addition and standardization of docstrings and  the addition of contributing  guide to docstrings in Contributing.md 
---

### CODE REVIEW

File: whittle/modules/embedding.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 
 
 class Embedding(torch.nn.Embedding):
+    """An extension of PyTorch's `torch.nn.Embedding` with support of sub-network dimensionality."""
```
Comment (rheasukthanker) [inline whittle/modules/embedding.py:9]: Maybe rephrase as "An extension of PyTorch's `torch.nn.Embedding` with support  to sub-sample weights corresponding to the sub-network dimensionality"

File: whittle/modules/embedding.py [hunk 2]
```diff
@@ -34,9 +36,11 @@ def __init__(
         self.sub_network_embedding_dim: int | None = embedding_dim
 
     def set_sub_network(self, sub_network_embedding_dim: int):
+        """Set the embedding dimensionality of the current sub-network."""
         self.sub_network_embedding_dim = sub_network_embedding_dim
 
     def reset_super_network(self):
+        """Reset the embedding dimensionality of the current sub-network to the original value."""
```
Comment (rheasukthanker) [inline whittle/modules/embedding.py:43]: replace original value -> super-network dimensionality

File: whittle/modules/layernorm.py [hunk 3]
```diff
@@ -5,6 +5,8 @@
 
 
 class LayerNorm(torch.nn.LayerNorm):
+    """An extension of PyTorch's `torch.nn.LayerNorm` with support of sub-network dimensionality."""
```
Comment (rheasukthanker) [inline whittle/modules/layernorm.py:8]: Rephrase similar to embedding suggestion

File: whittle/modules/layernorm.py [hunk 4]
```diff
@@ -13,9 +15,11 @@ def __init__(self, in_features: int, eps: float = 1e-5):
         self.sub_network_in_features = self.in_features
 
     def set_sub_network(self, sub_network_in_features: int):
+        """Set the input dimensionality of the current sub-network."""
         self.sub_network_in_features = sub_network_in_features
 
     def reset_super_network(self):
+        """Reset the input dimensionality of the current sub-network to the original value."""
```
Comment (rheasukthanker) [inline whittle/modules/layernorm.py:22]: Rephrase similar to embedding suggestion

File: whittle/modules/linear.py [hunk 5]
```diff
@@ -6,6 +6,8 @@
 
 
 class Linear(nn.Linear):
+    """An extension of PyTorch's `torch.nn.Linear` with support of sub-network dimensionality."""
```
Comment (rheasukthanker) [inline whittle/modules/linear.py:9]: """An extension of PyTorch's `torch.nn.Linear` with flexible input and output dimensionality corresponding to sub-network"""

File: whittle/modules/linear.py [hunk 6]
```diff
@@ -25,10 +26,12 @@ def __init__(
     def set_sub_network(
         self, sub_network_in_features: int, sub_network_out_features: int
     ):
+        """Set the linear transformation dimensions of the current sub-network."""
         self.sub_network_in_features = sub_network_in_features
         self.sub_network_out_features = sub_network_out_features
 
     def reset_super_network(self):
+        """Reset the linear transformation dimensions of the current sub-network to the original value."""
```
Comment (rheasukthanker) [inline whittle/modules/linear.py:34]: Rephrase similar to embedding suggestion. "original value" is ambiguous here

File: whittle/sampling/random_sampler.py [hunk 7]
```diff
@@ -1,23 +1,45 @@
...
 
+from typing import Any
+
 import numpy as np
 from syne_tune.config_space import Categorical, Domain
 
 
 class RandomSampler:
+    """
+    RandomSampler samples configurations from a given configuration space using a random state.
```
Comment (rheasukthanker) [inline whittle/sampling/random_sampler.py:12]: replace configuration space -> search space

File: whittle/sampling/random_sampler.py [hunk 8]
```diff
@@ -1,23 +1,45 @@
...
+
+    Args:
+        config_space: The configuration space from which to sample.
```
Comment (rheasukthanker) [inline whittle/sampling/random_sampler.py:15]: configuration space->search space

File: whittle/sampling/random_sampler.py [hunk 9]
```diff
@@ -1,23 +1,45 @@
...
+    """
+
     def __init__(self, config_space: dict, seed: int | None = None):
         self.config_space = config_space
         self.rng = np.random.RandomState(seed)
 
-    def sample(self):
+    def sample(self) -> dict[str, Any]:
+        """
+        Gets the smallest sub-network configuration from the configuration space.
```
Comment (rheasukthanker) [inline whittle/sampling/random_sampler.py:25]: Same as above

File: whittle/sampling/random_sampler.py [hunk 10]
```diff
@@ -1,23 +1,45 @@
...
         config = {}
         for hp_name, hparam in self.config_space.items():
             if isinstance(hparam, Domain):
                 config[hp_name] = hparam.sample(random_state=self.rng)
         return config
 
-    def get_smallest_sub_network(self):
+    def get_smallest_sub_network(self) -> dict[str, Any]:
+        """
+        Gets the smallest sub-network configuration from the configuration space.
```
Comment (rheasukthanker) [inline whittle/sampling/random_sampler.py:38]: Same as above

File: whittle/sampling/random_sampler.py [hunk 11]
```diff
@@ -1,23 +1,45 @@
...
+
+        Returns:
+            dict: The smallest sub-network configuration.
```
Comment (rheasukthanker) [inline whittle/sampling/random_sampler.py:41]: same as above

File: whittle/sampling/random_sampler.py [hunk 12]
```diff
@@ -33,7 +55,14 @@ def get_smallest_sub_network(self):
                     config[k] = v.lower
         return config
 
-    def get_largest_sub_network(self):
+    def get_largest_sub_network(self) -> dict[str, Any]:
+        """
+        gets the largest sub-network configuration from the configuration space.
```
Comment (rheasukthanker) [inline whittle/sampling/random_sampler.py:60]: same as above

File: whittle/search/ask_tell_scheduler.py
Comment (rheasukthanker) [inline whittle/search/ask_tell_scheduler.py:1]: I am not sure if we use this file anywhere currently. If not could you create an issue to remove it?

File: whittle/search/multi_objective.py [hunk 13]
```diff
@@ -5,10 +5,15 @@
 
 
 def get_pareto_optimal(costs: np.ndarray) -> npt.NDArray[np.bool_]:
-    """Find the pareto-optimal point.
+    """
+    Find the pareto-optimal point.
+
+    Args:
+        costs: (n_points, m_cost_values) array
```
Comment (rheasukthanker) [inline whittle/search/multi_objective.py:12]: It would be better to be more descriptive here, what do costs indicate? should they be increasing or decreasing?

Comment (zeqri) [reply whittle/search/multi_objective.py:12]: cost is an indication of the validation loss if I am not mistaken so they should be decreasing right?

File: whittle/models/gpt/blocks/causal_self_attention.py [hunk 14]
```diff
@@ -11,6 +11,8 @@
 
 
 class CausalSelfAttention(nn.Module):
+    """Extension of litgpt's `litgpt.model.CausalSelfAttention` with support to adapt to sub-network dimensionality."""
```
Comment (rheasukthanker) [inline whittle/models/gpt/blocks/causal_self_attention.py:14]: this seems too general, we should be more specific on details here

**Review by rheasukthanker — CHANGES_REQUESTED** [review]
Could we follow the format in the following examples? Currently doc strings are quite in-consistent 
1. https://github.com/Lightning-AI/litgpt/blob/main/litgpt/model.py#L440
2. https://github.com/Lightning-AI/litgpt/blob/main/litgpt/lora.py#L66
3. https://github.com/Lightning-AI/litgpt/blob/main/litgpt/lora.py#L102
4. https://github.com/Lightning-AI/litgpt/blob/main/litgpt/lora.py#L192
5. https://github.com/Lightning-AI/litgpt/blob/main/litgpt/lora.py#L403

File: whittle/models/gpt/extract.py [hunk 15]
```diff
@@ -8,9 +9,24 @@
...
+    Copies relevant layers, weights, and configurations from the full model into a sub-network model.
+
+    Args:
+        model: The original, full GPT model from which the sub-network is extracted.
+        sub_network_config: Configuration object for the sub-network, containing the necessary
+                                     architecture specifications such as embedding size, number of heads,
+                                     and number of layers.
 
+    Returns:
+        GPT: A new sub-network model instance, initialized with parameters extracted from the original model.
```
Comment (timurcarstensen) [inline whittle/models/gpt/extract.py:27]:
```suggestion
        A new sub-network model instance, initialized with parameters extracted from the original model.
```

I would encourage to not include the type of the return or args in the docstring, if the type in the function signature changes you'll have to update two things which can often be overlooked by accident. Only adding the types in the function signature is therefore more maintainable :)

File: whittle/sampling/random_sampler.py [hunk 16]
```diff
@@ -1,23 +1,45 @@
...
         self.config_space = config_space
         self.rng = np.random.RandomState(seed)
 
-    def sample(self):
+    def sample(self) -> dict[str, Any]:
+        """
+        Gets the smallest sub-network configuration from the search space.
+
+        Returns:
+            dict: The smallest sub-network configuration.
```
Comment (timurcarstensen) [inline whittle/sampling/random_sampler.py:28]:
```suggestion
            The smallest sub-network configuration.
```

File: whittle/search/local_search.py [hunk 17]
```diff
@@ -27,21 +27,21 @@ class PopulationElement:
...
-    :param random_seed: Random seed, optional
-    :param kwargs: Additional arguments to
-        :class:`~syne_tune.optimizer.schedulers.FIFOScheduler`
+       for ``kwargs["search_options"]`` parameters.
+
+       Args:
+           config_space: Configuration space for evaluation function
+           metric: Name of metric to optimize
+           population_size: See
+               :class:`~syne_tune.optimizer.schedulers.searchers.RegularizedEvolution`.
```
Comment (timurcarstensen) [inline whittle/search/local_search.py:37]: Do these links to different classes actually work with our mkdocs configuration?

File: whittle/search/search.py [hunk 18]
```diff
@@ -16,19 +17,26 @@ def multi_objective_search(
...
-    :param objective_kwargs: the keyword arguments for the objective function.
-    :param num_samples: the number of samples to take.
-    :param seed: the random seed.
-    :return: the results of the search.
+) -> dict[str, Any]:
     """
+    Search for the Pareto-optimal sub-networks using the specified strategy.
+
+    Args:
+        objective (callable): The objective function to optimize.
```
Comment (timurcarstensen) [inline whittle/search/search.py:25]:
```suggestion
        objective: The objective function to optimize.
```

File: whittle/search/search.py [hunk 19]
```diff
@@ -16,19 +17,26 @@ def multi_objective_search(
...
+            Defaults to "random_search".
+        num_samples: The number of samples to evaluate.
+            Defaults to 100.
+        objective_kwargs: Keyword arguments for the objective function.
+            Defaults to None.
+        seed: The random seed for reproducibility.
+            Defaults to None.
 
+    Returns:
+        dict: The results of the search, including Pareto-optimal solutions.
```
Comment (timurcarstensen) [inline whittle/search/search.py:37]:
```suggestion
        The results of the search, including Pareto-optimal solutions.
```

File: whittle/search/local_search.py (hunk 17 above)
Comment (zeqri) [reply whittle/search/local_search.py:37]: they are not shown as links

Comment (zeqri) [reply whittle/search/local_search.py:37]: I didn't write this docstring initially, I just reformulated how it was written.

Comment (timurcarstensen) [reply whittle/search/local_search.py:37]: ping @aaronkl

Comment (aaronkl) [reply whittle/search/local_search.py:37]: we can remove these links

File: whittle/models/gpt/blocks/causal_self_attention.py [hunk 20]
```diff
@@ -73,6 +76,7 @@ def set_sub_network(
             self.sub_attention_scaler = self.config.attention_scores_scalar
 
     def reset_super_network(self):
+        """Resets the dimensionality of the current to the super-network dimensionality."""
```
Comment (timurcarstensen) [inline whittle/models/gpt/blocks/causal_self_attention.py:79]: > [...] of the current [...]

Did you mean of the current model/sub-network? :)

File: whittle/models/gpt/blocks/mlp.py [hunk 21]
```diff
@@ -24,6 +26,7 @@ def __init__(self, config: Config) -> None:
     def set_sub_network(
         self, sub_network_n_embd: int, sub_network_intermediate_size: int
     ):
+        """Set the input dimensionality of the current sub-network."""
```
Comment (timurcarstensen) [inline whittle/models/gpt/blocks/mlp.py:29]: I think input is out of place here, no? @rheasukthanker

File: whittle/models/gpt/blocks/mlp.py [hunk 22]
```diff
@@ -35,6 +38,7 @@ def set_sub_network(
         )
 
     def reset_super_network(self):
+        """Reset the input dimensionality of the current sub-network to the super-network dimensionality."""
```
Comment (timurcarstensen) [inline whittle/models/gpt/blocks/mlp.py:41]: same as above (maybe I'm wrong tho, feel free to correct)

File: whittle/models/gpt/extract.py [hunk 23]
```diff
@@ -8,9 +9,24 @@
...
+def extract_sub_network(model: GPT, sub_network_config: Config) -> GPT:
+    """
+    Extracts a sub-network from a given model based on the specified sub-network configuration.
+    Copies relevant layers, weights, and configurations from the full model into a sub-network model.
+
+    Args:
+        model: The original, full GPT model from which the sub-network is extracted.
+        sub_network_config: Configuration object for the sub-network, containing the necessary
+                                     architecture specifications such as embedding size, number of heads,
+                                     and number of layers.
```
Comment (timurcarstensen) [inline whittle/models/gpt/extract.py:24]: The line breaks seem a little weird here, maybe follow what eddie does [here in amltk](https://github.com/automl/amltk/blob/5221d80fb4fec92f2d6f2fc73fce4eb16139c09f/src/amltk/pipeline/components.py#L169)

**Review by timurcarstensen — COMMENTED** [review]
Generally looking really good! 😄 I left some small comments but after those are resolved feel free to merge this pr

File: whittle/models/gpt/blocks/mlp.py (hunk 22 above)
Comment (zeqri) [reply whittle/models/gpt/blocks/mlp.py:41]: This one mentions the current sub-network, it is already correct isn't it?

File: whittle/models/gpt/blocks/causal_self_attention.py (hunk 20 above)
Comment (timurcarstensen) [inline whittle/models/gpt/blocks/causal_self_attention.py:79]:
```suggestion
        """Resets the dimensionality of the current sub-network to the super-network dimensionality."""
```

File: whittle/models/gpt/blocks/mlp.py (hunk 22 above)
Comment (timurcarstensen) [reply whittle/models/gpt/blocks/mlp.py:41]: When I was referring to the above comment I meant the `input` being out of place :D This function does not just reset the input size but also the intermediate so describing it as just resetting the input size would be incorrect, right?

Comment (zeqri) [reply whittle/models/gpt/blocks/mlp.py:41]: I agree
